python/kvikio/_lib/libnvcomp_ll.pyx

```diff
@@ -333,8 +430,61 @@ class nvCompBatchAlgorithmLZ4(nvCompBatchAlgorithm):
             &max_compressed_bytes
         )
 
+        # Add header size, if needed.
+        if err == nvcompStatus_t.nvcompSuccess and self.has_header:
+            max_compressed_bytes += self.HEADER_SIZE_BYTES
+
         return (err, max_compressed_bytes)
 
+    def compress(
+        self,
+        uncomp_chunks,
+        uncomp_chunk_sizes,
+        size_t max_uncomp_chunk_bytes,
+        size_t batch_size,
+        temp_buf,
+        comp_chunks,
+        comp_chunk_sizes,
+        stream,
+    ):
+        if self.has_header:
+            # If there is a header, we need to:
+            # 1. Copy the uncompressed chunk size to the compressed chunk header.
+            # 2. Update target pointers in comp_chunks to skip the header portion,
+            # which is not compressed.
+            #
+            # Get the base pointers to sizes.
+            psize = to_ptr(uncomp_chunk_sizes)
+            for i in range(batch_size):
+                # Copy the original data size to the header.
+                memcpyAsync(
+                    <uintptr_t>comp_chunks[i],
+                    psize,
+                    sizeof(uint32_t),
+                    cudaMemcpyKind.cudaMemcpyDeviceToDevice,
+                    stream.ptr
+                )
```

Comment:
Could perhaps also use cudamemcpy2dasync here too (see below).

Reply:
In this case, `comp_chunks` is currently a host array which keeps pointers to every chunk. Since those chunks were allocated originally as a 2D array, we potentially could pass that one here instead. But then we still need access to each chunk pointer separately to update it before passing to nvCOMP compress and then resetting it back. 
Possibly we could do that by carefully crafting and passing host/device arrays, but I'd like to keep it as is for now. I suggest we create a separate issue/task for that.